rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+The changes we propose will make it easier for clients to start using Teleport for accessing Kubernetes clusters.
+
+## Scope
+
+This RFD is focused only on Kubernetes resources related to RBAC functionality, other types of resources are out of scope
+and we will not allow them to be provisioned. Although other resources might be added in the future.
+
+## Details
+
+### Improving Day 1 experience - Adding capability to manage Kubernetes RBAC from Teleport. 
+
+Currently users rely on `kubernetes_groups` and/or `kubernetes_users` fields of the role definition when setting up
+access for Kubernetes. This requires roles/bindings to be created by the client in Kubernetes cluster independently from the 
+RBAC setup in the Teleport. These roles/bindings also should be kept in sync with the Teleport RBAC side,
+which creates additional hurdles for clients, especially the new ones, when using Kubernetes access. 
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
+pair on the Kubernetes cluster, otherwise Role/Binding pair will be created in the each listed namespace. 
+
+Subject for the bindings will be a group named `teleport:*RoleName*`, e.g. for a Teleport role `kube-role` the resulting 
+group name will be `teleport:kube-role`. This will allow Kube service to setup correct impersonation group headers based on the roles available to the user.
+
+If `kubernetes_permissions` field is set in the `allow` part of the Teleport role, fields `kubernetes_resources`, `kubernetes_users` and `kubernetes_groups` 
```

Comment:
Why would it be a blocker? If I have 200 roles that use kube_users, kube_groups, kube_resources, why can't I mix them given that kube permissions are always additions.

Reply:
To lessen potential confusion and give us an option to enable it later, if we allow it from the beginning and then find some problems we can't roll it back. If you're existing user with 200 roles you already have complicated system and probably some way of herding your Kubernetes RBAC, feels like mixing approaches might be not the cleanest way. You still can create completely new roles with `kubernetes_permissions` or convert your existing ones. For newer users it should be less confusing though. And as I said, we can add this functionality later once we see how it's going.